File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.kt
class Solution {
    fun validPath(n: Int, edges: Array<IntArray>, source: Int, destination: Int): Boolean {
        val g = Array(n) { arrayListOf<Int>() }
        edges.forEach { (u, v) ->
            g[u].add(v)
            g[v].add(u)
        }
        val passed = hashSetOf(source)
        var level = listOf(source)
        while (level.isNotEmpty()) {
            if (level.any(destination::equals)) return true 
            level = level.flatMap(g::get).filterNot(passed::contains)
            passed.addAll(level)
        }
        return false
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n, edges, s, d) -> bool:
        g = [[] for _ in range(n)]
        for e in edges:
            g[e[0]].append(e[1])
            g[e[1]].append(e[0])
        level, passed = [s], {s}
        while level:
            if any(n == d for n in level): return True
            level = [c for n in level for c in g[n] if c not in passed]
            for n in level: passed.add(n)
        return False
        
        